### PR TITLE
stringutil: improve DoMatch performance

### PR DESCRIPTION
**Pull request by TennyZhuang**
### What problem does this PR solve? 

`DoMatch` has bad performance on special case.

related to https://github.com/tikv/tikv/pull/5866

### What is changed and how it works?

Optimize the DoMatch algorithm by https://research.swtch.com/glob

Backtrace only back to last `patAny`

All characters will be traversed at most 2 times.

### Check List 

Tests 

I add a new special case benchmark test.

Before optimization:

```
PASS: string_util_test.go:128: testStringUtilSuite.TestIsExactMatch	0.000s
PASS: string_util_test.go:80: testStringUtilSuite.TestPatternMatch	0.000s
PASS: string_util_test.go:33: testStringUtilSuite.TestUnquote	0.000s
OK: 3 passed
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/util/stringutil
BenchmarkMatchSpecial-8   	       3	 426341220 ns/op
PASS
ok  	github.com/pingcap/tidb/util/stringutil	2.604s

```

After optimization:

```
PASS: string_util_test.go:128: testStringUtilSuite.TestIsExactMatch	0.000s
PASS: string_util_test.go:80: testStringUtilSuite.TestPatternMatch	0.000s
PASS: string_util_test.go:33: testStringUtilSuite.TestUnquote	0.000s
OK: 3 passed
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/util/stringutil
BenchmarkMatchSpecial-8   	 8777878	       135 ns/op
PASS
ok  	github.com/pingcap/tidb/util/stringutil	1.336s

```

 - Unit test

Code changes

Side effects

Related changes

Release note

 - Optimize DoMatch performance

### CODE REVIEW

**Comment by TennyZhuang** [comment]
@breeswish PTAL

**Comment by zz-jason** [comment]
@TennyZhuang It's better to add some benchmarks for the performance improvement. Could you add one in a test file to tell the code reviewers about how much performance does this PR improved?

**Comment by TennyZhuang** [comment]
@zz-jason I have add a benchmark, and add my result in description.

**Comment by TennyZhuang** [comment]
ping

File: util/stringutil/string_util.go [hunk 1]
```diff
@@ -206,35 +206,45 @@ func matchByteCI(a, b byte) bool {
...
+				}
+			case patOne:
+				if sIdx < len(str) {
+					pIdx++
+					sIdx++
+					continue
 				}
-				sIdx++
+			case patAny:
+				// Try to match at nx.
```
Comment (wshwsh12) [inline util/stringutil/string_util.go]:
```suggestion
				// Try to match at sIdx.
```

File: util/stringutil/string_util.go [hunk 2]
```diff
@@ -206,35 +206,45 @@ func matchByteCI(a, b byte) bool {
...
+				// If that doesn't work out,
+				// restart at nx+1 next.
```
Comment (wshwsh12) [inline util/stringutil/string_util.go]:
```suggestion
				// restart at sIdx+1 next.
```

File: util/stringutil/string_util_test.go [hunk 3]
```diff
@@ -152,3 +152,19 @@ func (s *testStringUtilSuite) TestIsExactMatch(c *C) {
...
+		escape  = byte('\\')
+	)
+
+	patChars, patTypes := CompilePattern(pattern, escape)
+
+	b.StartTimer()
+	match := DoMatch(target, patChars, patTypes)
+	if match {
+		b.Fatal("Unmatch expected.")
+	}
```
Comment (wshwsh12) [inline util/stringutil/string_util_test.go]: The benchmark usage is wrong. Have a try using following code. And address the benchmark result in the description.
```
for i := 0;i < b.N;i++ {
    match := DoMatch(target, patChars, patTypes)
    if match {
        b.Fatal("Unmatch expected.")
    }
}
```

**Comment by TennyZhuang** [comment]
@wshwsh12 PTAL

**Comment by TennyZhuang** [comment]
@wshwsh12 I have updated my benchmark result, PTAL

**Comment by TennyZhuang** [comment]
And the corresponding optimization is merged in https://github.com/tikv/tikv/pull/5866

File: util/stringutil/string_util_test.go [hunk 4]
```diff
@@ -152,3 +152,21 @@ func (s *testStringUtilSuite) TestIsExactMatch(c *C) {
...
+func BenchmarkMatchSpecial(b *testing.B) {
+	var (
+		pattern = `a%a%a%a%a%a%a%a%b`
+		target  = `aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa`
+		escape  = byte('\\')
+	)
+
+	patChars, patTypes := CompilePattern(pattern, escape)
+
+	b.StartTimer()
```
Comment (SunRunAway) [inline util/stringutil/string_util_test.go]:
```suggestion
	b.ResetTimer()
```
The benchmarking framework starts the timer automatically, see https://golang.org/pkg/testing/#B.StartTimer .
So we should call b.StopTimer at the begin of `BenchmarkMatchSpecial`, or just call `b.ResetTimer`.

**Comment by TennyZhuang** [comment]
@SunRunAway ping

**Comment by TennyZhuang** [comment]
cm

**Comment by TennyZhuang** [comment]
emmm

cla is blocked(

**Pull request merged**